app/lib/pdf/finance/invoice_report_data.rb

```diff
@@ -52,23 +52,22 @@ def plan
   end
 
   def logo
-    @logo_stream ||= URI.open(logo_url(:invoice))
+    @logo_file ||= logo_file(:invoice)
   rescue StandardError => exception
-    # TODO: - uncomment complete exception!
-    # rescue OpenURI::HTTPError => e
     Rails.logger.error "Failed to retrieve logo from: #{exception.message}"
     nil
   end
 
   # Depending on the storage option for the attachment, retrieve either the full path
-  # of the local file, or the URL of the file in S3
-  def logo_url(args)
+  # of the local file, or the URL of the file in S3, and read the file differently
+  def logo_file(style)
     attachment = @invoice.provider_account.profile.logo
     case attachment.options[:storage].to_sym
     when :filesystem
-      attachment.path(args)
+      # read as binary file
+      File.open(attachment.path(style), 'rb')
```

Comment:
Smells like a paperclip bug. If `attachment.path(args)` is a `Pathname`, then maybe `attachment.path(args).open('rb')`.

Reply:
It's a String. Are there any benefits in using `Pathname#open` rather than `File.open`?

It's not really a paperclip bug, it's about how prawn [handles the images](https://github.com/3scale/prawn/blob/0.5.1-3scale/lib/prawn/images.rb#L64-L69). `open("/local/path")` will create an IO object with `.read`, but it is read as UTF-8, not binary (`ASCII-8BIT`).

Reply:
God it. My concern is that we open a file descriptor here, then is it closed by prawn? If prawn opens something, I assume it will close it after. But if we open, then who closes it?

Reply:
Yes, you're totally right! prawn does not close them.
It actually has always been like that - we were just passing the open file descriptor (see the original `open(@invoice.provider_account.profile.logo.url(:invoice))`)

I believe they will be eventually collected by GC, but we should really be closing them.
Thanks for spotting! I will need to refactor this more significantly to ensure we're closing them.

For files we can indeed just pass the file full path, and prawn will open-read-close the file (`File.binread`).
But for S3-stored files I guess we still need to pass an IO object with a `read` method, so prawn get [get the content](https://github.com/3scale/prawn/blob/0.5.1-3scale/lib/prawn/images.rb#L65).

Reply:
How about use `attachment#copy_to_local_file` and pass as a string?

Reply:
But we will need to keep a reference to the file handle until it is used. Otherwise the tempfile can be GCed and removed :/

So not sure it's gonna be easier this way.